From: "Matteo F. Vescovi" <mfv@debian.org>
Date: Mon, 23 Jan 2017 14:18:06 +0100
Subject: Fix_typo_in_help_entry

---
 bs1770gain/bs1770gain.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bs1770gain/bs1770gain.c b/bs1770gain/bs1770gain.c
index ffae8aa..ab5e721 100755
--- a/bs1770gain/bs1770gain.c
+++ b/bs1770gain/bs1770gain.c
@@ -92,7 +92,7 @@ void bs1770gain_usage(char **argv, int code)
   /////////////////////////////////////////////////////////////////////////////
   FPRINTF(stderr," -i,--integrated:  calculate integrated loudness\n");
   FPRINTF(stderr," -s,--shorterm:  calculate maximum shortterm loudness\n");
-  FPRINTF(stderr," -m,--momentary:  claculate maximum momentary loudness\n");
+  FPRINTF(stderr," -m,--momentary:  calculate maximum momentary loudness\n");
   FPRINTF(stderr," -r,--range:  calculate loudness range\n");
   FPRINTF(stderr," -p,--samplepeak:  calculate maximum sample peak\n");
   FPRINTF(stderr," -t,--truepeak:  calculate maximum true peak\n");
